test/e2e/upgrade_test.go

```diff
@@ -0,0 +1,408 @@
+			},
+			Spec: metal3api.BareMetalHostSpec{
+				Online: true,
+				BMC: metal3api.BMCDetails{
+					Address:         bmcAddress,
+					CredentialsName: "bmc-credentials",
+				},
+				BootMode:       metal3api.Legacy,
+				BootMACAddress: bootMacAddress,
+			},
+		}
+		err = upgradeClusterProxy.GetClient().Create(ctx, &bmh)
+		Expect(err).NotTo(HaveOccurred())
+
+		By("Waiting for the BMH to become available")
+		WaitForBmhInProvisioningState(ctx, WaitForBmhInProvisioningStateInput{
+			Client: upgradeClusterProxy.GetClient(),
+			Bmh:    bmh,
+			State:  metal3api.StateAvailable,
+		}, e2eConfig.GetIntervals(specName, "wait-available")...)
+
+		By("Upgrading BMO deployment")
+		clientSet := upgradeClusterProxy.GetClientSet()
+		deploy, err := clientSet.AppsV1().Deployments(bmoNamespace).Get(ctx, bmoDeployName, metav1.GetOptions{})
+		Expect(err).NotTo(HaveOccurred())
+		manifest, err := buildKustomizeManifest(bmoKustomization)
+		Expect(err).NotTo(HaveOccurred())
+		err = upgradeClusterProxy.Apply(ctx, manifest)
+		Expect(err).NotTo(HaveOccurred())
+		By("Waiting for BMO update to rollout")
+		Eventually(func() bool {
+			return DeploymentRolledOut(ctx, upgradeClusterProxy, bmoDeployName, bmoNamespace, deploy.Status.ObservedGeneration+1)
+		},
+			e2eConfig.GetIntervals("default", "wait-deployment")...,
+		).Should(BeTrue())
+		By("Patching the BMH to test provisioning")
+		helper, err := patch.NewHelper(&bmh, upgradeClusterProxy.GetClient())
+		Expect(err).NotTo(HaveOccurred())
+		bmh.ObjectMeta.Annotations[metal3api.HardwareDetailsAnnotation] = hardwareDetails
+		bmh.Spec.Image = &metal3api.Image{
+			URL:      e2eConfig.GetVariable("IMAGE_URL"),
+			Checksum: e2eConfig.GetVariable("IMAGE_CHECKSUM"),
+		}
+		bmh.Spec.RootDeviceHints = &metal3api.RootDeviceHints{
+			DeviceName: "/dev/vda",
+		}
+		Expect(helper.Patch(ctx, &bmh)).To(Succeed())
+
+		By("Waiting for the BMH to be in provisioning state")
+		WaitForBmhInProvisioningState(ctx, WaitForBmhInProvisioningStateInput{
+			Client: upgradeClusterProxy.GetClient(),
+			Bmh:    bmh,
+			State:  metal3api.StateProvisioning,
+		}, e2eConfig.GetIntervals(specName, "wait-provisioning")...)
+
+		By("Waiting for the BMH to become provisioned")
+		WaitForBmhInProvisioningState(ctx, WaitForBmhInProvisioningStateInput{
+			Client: upgradeClusterProxy.GetClient(),
+			Bmh:    bmh,
+			State:  metal3api.StateProvisioned,
+		}, e2eConfig.GetIntervals(specName, "wait-provisioned")...)
+	})
+
+	AfterEach(func() {
+		cancelWatches()
+		if !skipCleanup {
+			if upgradeClusterProxy != nil {
+				upgradeClusterProxy.Dispose(ctx)
+			}
+			if upgradeClusterProvider != nil {
+				upgradeClusterProvider.Dispose(ctx)
+			}
+		}
+		if useExistingCluster {
+			k8sContextName := e2eConfig.GetVariable("KUBECONFIG_DEFAULT_CONTEXT_NAME")
+			By(fmt.Sprintf("Switch back to kubeconfig context: %s", k8sContextName))
+			err := SwitchKubeconfigContext(kubeconfigPath, k8sContextName)
+			Expect(err).NotTo(HaveOccurred())
```

Comment:
Shouldn't we always switch the context here, not just when using existing cluster?
I think it would be even better to check what context was used before the test and switch back to that, instead of switching to what the variable is. If the variable is not set this breaks.